Use ieee intrinsics to avoid trapping ieee_inexact signals

peterdschwartz opened the pull request:
Use ieee intrinsics to avoid trapping ieee_inexact signals potentially due to a compiler bug in nvhpc.
ELM is halting on `ieee_inexact` values when `Ktrap=fp` mode is set. 

Fixes #7897 

[BFB]

peterdschwartz commented:
going to perform sanity checks on chrysalis

peterdschwartz commented:
Adding ifdefs is unnecessary:  trapping `inexact` is always off for all compilers and machines, including `pm-cpu_nvidia`.  This workaround appears necessary due to a bug with nvidia's math runtime.  If someone wanted to turn on traps for inexact, the correct way would be to use the ieee intrinsics around the code of interest rather than compile flags.

peterdschwartz commented on components/elm/src/biogeophys/SoilStateType.F90:
```diff
@@ -468,10 +469,13 @@ subroutine InitCold(this, bounds)
-       !    Variable ZSOI not found, use the ELM parameters.
+       if (ieee_support_halting(ieee_inexact)) then
+          call ieee_set_flag(ieee_all,.false.)
+          call ieee_set_halting_mode(ieee_inexact, .false.)
```
this is essentially a no-op except for this nvida on perlmutter, which i can only hazard is a bug in the compiler or runtime.

peterdschwartz replied:
it's even false with nvidia !  From the nvidia docs our current flag only captures `inv, divz, ovf`.  iexact is really never used because all transcendental functions are inexact.  I have verified this by calling `ieee_get_halting_mode(ieee_inexact, halt)`  which is `.false.` wherever i call it.
```
The special flag -Ktrap=none is used to preserve FPEs during compilation without unmasking any of them at runtime.

The inv, divz, and ovf flags are often the most interesting, as they signify abnormal floating-point behavior in the program. These can be enabled with the useful shorthand -Ktrap=fp.
```

it seems to be a bug in the code gen or AVX2 exp runtime.  Found a similar issue of random inexact being tripped in 25.9 https://forums.developer.nvidia.com/t/nvfortran-25-9-spurious-floating-point-exception/346604

peterdschwartz commented:
@ndkeen 
`o.gnu-trap.txt:     Flags  after     exp:  inv=  F      divz=  F     ovf=  F     unf=  F     inx=  T` 

this shows that the only ieee flag being set from the `exp` is the `ieee_inexact` flag.  This proves that the program isn't catching an `ieee_invalid`, `ieee_overflow`, or any of the others.  Note that every call to `exp` (and any other math transcendental function) sets the `ieee_inexact` flag, which is why compilers do not trap it by default and some don't even allow for trapping it on certain architectures.

You did not show the output from this line:
```fortran
   call ieee_get_halting_mode(ieee_inexact, halt_inexact)
   print *, "halt_inexact AT THE TOP: ",halt_inexact
```
But if you did, that would tell you if the program is set to halt (ie trap) the `ieee_inexact`.  You should find that none of those compilers with those flags do, or if they did, the program would halt with a `SIGFPE`

ndkeen commented:
```
muller-login02% grep halt o.* | column -t
o.gnu-trap.txt:     halt_inexact  AT  THE  TOP:  F
o.gnu.txt:          halt_inexact  AT  THE  TOP:  F
o.intel-trap.txt:   halt_inexact  AT  THE  TOP:  F
o.intel.txt:        halt_inexact  AT  THE  TOP:  F
o.nvida.txt:        halt_inexact  AT  THE  TOP:  F
o.nvidia-trap.txt:  halt_inexact  AT  THE  TOP:  F
```
you are correct, none have it.

But while its odd that inexact is showing T, it's not clear why nvidia would behave differently than others?
I mean, can we modify this test to show?

peterdschwartz commented:
It's not odd to show inexact = T after an exp.  Based on this thread https://forums.developer.nvidia.com/t/nvfortran-25-9-spurious-floating-point-exception/346604

Someone was able to make a small reproducer with 25.9 and with -O2 flag.  It may be worth simply copying that code and seeing if we get the same results.  But Matt Colgrove confirms that it shouldn't be happening and is a compiler bug.

ndkeen commented:
OK, using that test code provided, I was able to get a signal. If I build with debug flags _and_ `-O2`, I can get a FP exception (only with nvidia compiler). However, with e3sm, we arent building with `-O2`.  Also, for test problem, I can get the error to go away (as suggested) with `-Mnoinline`, which is actually a reasonable thing to add to our e3sm DEBUG flags. I tried this, but I get the error (although different location).  I then see that certain libs (outside of e3sm.bldlog) are for some reason not using all of the fortran debug flags specified. When I add them to be consistent across all fortran builds, I still get the error.

```
  0: Error: floating point exception, floating point invalid operation
  0:    rax 0x0000000000000080, rbx 0x00007ffdcc751cc0, rcx 0x0000000000000007
  0:    rdx 0x0000000000000000, rsp 0x00007ffdcc751b10, rbp 0x0000000000000000
  0:    rsi 0x0000000012bb0e98, rdi 0x0000000012bb09c0, r8  0x0000000012bb0e60
  0:    r9  0x0000000000000003, r10 0x0000000012a31010, r11 0x0000000000000006
  0:    r12 0x3ff0000000000000, r13 0x00007ffdcc752750, r14 0x00007ffdcc751b58
  0:    r15 0x00007ffdcc751b70
  0:   /lib64/libpthread.so.0(+0x16910) [0x1499ad66c910]
  0:   /opt/cray/pe/hdf5-parallel/1.14.3.1/nvidia/23.3/lib/libhdf5_parallel_nvidia.so.310(H5T__init_native_float_types+0xc34) [0x1499ac8bbbb4]
  0:   /opt/cray/pe/hdf5-parallel/1.14.3.1/nvidia/23.3/lib/libhdf5_parallel_nvidia.so.310(H5T_init+0x1a9) [0x1499ac80ef69]
  0:   /opt/cray/pe/hdf5-parallel/1.14.3.1/nvidia/23.3/lib/libhdf5_parallel_nvidia.so.310(H5VL_init_phase2+0x15d) [0x1499ac8de1dd]
  0:   /opt/cray/pe/hdf5-parallel/1.14.3.1/nvidia/23.3/lib/libhdf5_parallel_nvidia.so.310(H5_init_library+0x350) [0x1499ac587f90]
  0:   /opt/cray/pe/hdf5-parallel/1.14.3.1/nvidia/23.3/lib/libhdf5_parallel_nvidia.so.310(H5Eset_auto2+0x42) [0x1499ac6479c2]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdf_parallel_nvidia.so.19(+0x9be1b) [0x1499acaabe1b]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdf_parallel_nvidia.so.19(nc4_hdf5_initialize+0xa) [0x1499acaaab8a]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdf_parallel_nvidia.so.19(NC_HDF5_initialize+0x23) [0x1499acab5f63]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdf_parallel_nvidia.so.19(nc_initialize+0x58) [0x1499aca377d8]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdf_parallel_nvidia.so.19(NC_open+0x103) [0x1499aca3bb43]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdf_parallel_nvidia.so.19(nc_open+0x25) [0x1499aca3ade5]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdff_parallel_nvidia.so.7(nf_open_+0x14c) [0x1499acc0d98c]
  0:   /opt/cray/pe/netcdf-hdf5parallel/4.9.0.13/nvidia/23.3/lib/libnetcdff_parallel_nvidia.so.7(netcdf_nf90_open_+0x154) [0x1499acc33e54]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(shr_stream_mod_shr_stream_getcalendar_: shr_stream_mod_shr_stream_\
getcalendar_ at /global/cfs/cdirs/e3sm/ndk/repos/c31-nov6/share/streams/shr_stream_mod.F90:1949) [0x25ef010]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(shr_stream_mod_shr_stream_init_: shr_stream_mod_shr_stream_init_ a\
t /global/cfs/cdirs/e3sm/ndk/repos/c31-nov6/share/streams/shr_stream_mod.F90:624) [0x25dfffb]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(shr_strdata_mod_shr_strdata_readnml_: shr_strdata_mod_shr_strdata_\
readnml_ at /global/cfs/cdirs/e3sm/ndk/repos/c31-nov6/share/streams/shr_strdata_mod.F90:1429) [0x25a67cb]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(datm_shr_mod_datm_shr_read_namelists_: datm_shr_mod_datm_shr_read_\
namelists_ at /global/cfs/cdirs/e3sm/ndk/repos/c31-nov6/components/data_comps/datm/src/datm_shr_mod.F90:147) [0xa5626f]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(atm_comp_mct_atm_init_mct_: atm_comp_mct_atm_init_mct_ at /global/\
cfs/cdirs/e3sm/ndk/repos/c31-nov6/components/data_comps/datm/src/atm_comp_mct.F90:153) [0xa36d34]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(component_mod_component_init_cc_: component_mod_component_init_cc_\
 at /global/cfs/cdirs/e3sm/ndk/repos/c31-nov6/driver-mct/main/component_mod.F90:259) [0x848b80]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(cime_comp_mod_cime_init_: cime_comp_mod_cime_init_ at /global/cfs/\
cdirs/e3sm/ndk/repos/c31-nov6/driver-mct/main/cime_comp_mod.F90:1518) [0x80f9b1]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(MAIN_: MAIN_ at /global/cfs/cdirs/e3sm/ndk/repos/c31-nov6/driver-m\
ct/main/cime_driver.F90:124) [0x845b0b]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(main+0x31) [0x8006f1]
  0:   /lib64/libc.so.6(__libc_start_main+0xef) [0x1499a643e1fd]
  0:   /mscratch/sd/n/ndk/e3sm_scratch/muller-cpu/ERS_D.f09_f09.IELM.muller-cpu_nvidia.elm-koch_snowflake.pr7867-allfortflagssame/bld/e3sm.exe(_start: _start at /home/abuild/rpmbuild/BUILD/glibc-2.31/csu/../sy\
sdeps/x86_64/start.S:122) [0x8005da]
srun: error: nid001003: task 0: Exited with exit code 127
```

This *might* be pointing to a place where we are reading in a file and perhaps could be a clue.


Just for completeness:

```
muller-login02% ./ieee-build.sh 
gnu
gnu with -ffpe-trap=invalid,zero,overflow
gnu with -O3 -ffpe-trap=invalid,zero,overflow
intel
intel with -fpe0
intel with -O3 -fpe0
nvidia
test-procs-mod.F90:
ieee-nvidia-inexact.F90:
nvidia with -O0 -g -Ktrap=fp -Mbounds -Kieee
test-procs-mod.F90:
ieee-nvidia-inexact.F90:
nvidia with -i4 -Mstack_arrays  -Mextend -byteswapio -Mflushz -Kieee -Mallocatable=03 -traceback  -O0 -g -Ktrap=fp -Mbounds -Kieee  -Mfree
test-procs-mod.F90:
ieee-nvidia-inexact.F90:
nvidia with -O2 -g -Ktrap=fp -Kieee
test-procs-mod.F90:
ieee-nvidia-inexact.F90:
test-procs-mod.F90:
ieee-nvidia-inexact.F90:
./ieee-build.sh: line 60: 2184875 Floating point exception./a.out > o.nvidia-opttrap.txt
nvidia with -O2 -g -Ktrap=fp -Mbounds -Kieee
test-procs-mod.F90:
ieee-nvidia-inexact.F90:
test-procs-mod.F90:
ieee-nvidia-inexact.F90:
2.493u 1.017s 0:04.02 87.0% 0pf+0w

With 

program test

   use, intrinsic :: ieee_arithmetic  ! use this with ieee_usual
   !use, intrinsic :: ieee_exceptions ! use this for IEEE_DIVIDE_BY_ZERO
   use procs
   implicit none
   logical :: halt_inexact, inv, divz, ovf, unf, inx
   integer :: j
   integer, parameter :: nlevgrnd=6
   real(8) :: zsoi(nlevgrnd), scalez, zecoeff
   double precision :: a(33),b

   scalez=0.25
   zecoeff=0.25

   call ieee_get_flag(ieee_invalid,         inv)
   call ieee_get_flag(ieee_divide_by_zero,  divz)
   call ieee_get_flag(ieee_overflow,        ovf)
   call ieee_get_flag(ieee_underflow,       unf)
   call ieee_get_flag(ieee_inexact,         inx)
   write(*,*) 'Flags initial: inv=',inv,' divz=',divz,' ovf=',ovf,' unf=',unf,' inx=',inx

   call ieee_get_halting_mode(ieee_inexact, halt_inexact)
   print *, "halt_inexact AT THE TOP: ",halt_inexact

   call ieee_set_flag(ieee_all, .false.)
   do j = 1, nlevgrnd
      zsoi(j) = scalez*(exp(zecoeff*(dble(j)-0.5))-1.0)    !node depths
      call ieee_get_flag(ieee_invalid,         inv)
      call ieee_get_flag(ieee_divide_by_zero,  divz)
      call ieee_get_flag(ieee_overflow,        ovf)
      call ieee_get_flag(ieee_underflow,       unf)
      call ieee_get_flag(ieee_inexact,         inx)
      write(*,*) 'Flags after exp: inv=',inv,' divz=',divz,' ovf=',ovf,' unf=',unf,' inx=',inx
   enddo

   a = &
      [0.3d0,7d11,7d9,4d0,1d0,1d3,1d0,0.2d0, &
      9d11,1d10,5d0,1d3,1d3,1d0,1d3,1d5,2d9, &
      5d0,1d0,8d-1,9d1,4d2,5d-1,5d-1,1d-1,8d6, &
      1d0,0d0,0d0,0d0,0d0,0d0,0d0]
   print*,'check 1'
   call flush(6)
   call sub1(a,b)
   print*,'check 2'
   call flush(6)
end program test

and the module as noted in the nvidia forum
```

peterdschwartz commented:
I think all signs point to it being an issue in the compiler mistakenly turning on halting on inexact rather than the codebase, and I don't think it would be worth the effort to try and figure out what part of code gen process is misbehaving. 

There is really **no** use case for trapping `ieee_inexact`, so I'm happy with the solution I gave:  it's clear, robust, and it ensures `ieee_inexact` is not trapped which is what we need.

If you want to keep exploring, go ahead but there's no reason to hold up this PR while you do that.  There are potentially real ERS bugs with nvidia that need to be investigated.

peterdschwartz commented:
> I want this to be wrapped with #ifdef CPRNVIDIA because that tells future developers that this was done just for the NVIDIA compiler and **protects us against other compilers developing some bug with these routines**. Also add a comment saying what version of the NVIDIA compiler had this issue.

My point is that it not having the ifdef would protect us if other compilers had similar bugs.  As it stands it is a no-op for other compilers.  All adding a macro is doing, is hindering the readability of the code since several compilers require them to be on the first column of text.

peterdschwartz commented:
I've tested it @ndkeen and gave you the results in slack -- feel free to try it yourself.  I'll push changes to have this in an if statement but I do not have the bandwidth to keep re-investigating this.

peterdschwartz commented:
@rljacob I added the ifdef to set a `using_nvhpc` parameter to avoid the readability issue and added a comment.  

I would like to propose creating  (if it doesn't already exist)  module like `E3SM/share/utils/shr_config_mod.F90` that sets parameters based on macros for easy use throughout the code base.  Then each component could make their own if they want for their specific compile flags.   

for example,  `E3SM/components/elm/src/util/elm_config_mod.F90` could set parameters based on PETSc or modal aerosols, etc...

ndkeen commented:
Using the branch noted above, with next (that includes the ZM build issue fix), here are the nvidia fails:
```
ERS.f19_g16.I1850ELM.muller-cpu_nvidia.elm-betr.ed00                                                                                      fail                                 RUN nodes=   1 mins=  0.6 state=    FAILED notes=
ERS_D.ne4pg2_oQU480.F2010.muller-cpu_nvidia.eam-hommexx.ed00                                                                              fail                   COMPARE_base_rest nodes=   1 mins= 14.3 state= COMPLETED notes=
ERS_D_Ld15.f45_g37.IELMFATES.muller-cpu_nvidia.elm-fates_cold_treedamage.ed00                                                             fail                   COMPARE_base_rest nodes=   1 mins=  2.6 state= COMPLETED notes=
ERS_Ld20.f45_f45.IELMFATES.muller-cpu_nvidia.elm-fates.ed00                                                                               fail                   COMPARE_base_rest nodes=   1 mins=  2.0 state= COMPLETED notes=
ERS_Ld30.f45_f45.IELMFATES.muller-cpu_nvidia.elm-fates_satphen.ed00                                                                       fail                   COMPARE_base_rest nodes=   1 mins=  2.2 state= COMPLETED notes=
ERS_Ld30.f45_g37.IELMFATES.muller-cpu_nvidia.elm-fates_cold_sizeagemort.ed00                                                              fail                   COMPARE_base_rest nodes=   1 mins=  2.7 state= COMPLETED notes=
ERS_Ln9.ne4pg2_ne4pg2.F2010-MMF1.muller-cpu_nvidia.eam-mmf_crmout.ed00                                                                    fail                                 RUN nodes=   1 mins=  1.4 state=    FAILED notes=e3sm.log has non-ascii text
SMS_Lh4.ne4_ne4.F2010-SCREAMv1.muller-cpu_nvidia.eamxx-output-preset-1--eamxx-fixer_debug_output.ed00                                     fail                         MODEL_BUILD nodes=   0 mins=  0.0 state=           notes=
SMS_Lh4.ne4pg2_ne4pg2.F2010-SCREAMv1.muller-cpu_nvidia.eamxx-output-preset-1--eamxx-prod.ed00                                             fail                         MODEL_BUILD nodes=   0 mins=  0.0 state=           notes=
```

the 2 eamxx fails are expected.  the first one is a runtime error i think i posted about somewhere. 
but thats 82 passing

rljacob commented:
Oh and the last run was 22nd so yes should see new results tomorow.
